### PR TITLE
feature(worker): take leases by priority

### PR DESCRIPTION
**Pull request by bencvdb**
Part of https://fivetran.atlassian.net/browse/RD-41797

This PR allows `ShardPrioritization` to prioritize anything that implements `PriorizeableShard`. Additionally, it allows users to configure the Worker to take advantage of this new functionality to prioritize `KinesisClientLease` in [`LeaseTaker#computeLeasesToTake`](https://github.com/fivetran/amazon-kinesis-client/blob/e8dcb0a8106e6e2a708abaf59ba11f2cddb7e6bd/src/main/java/com/fivetran/external/com/amazonaws/services/kinesis/leases/impl/LeaseTaker.java#L381)  instead of selecting them randomly. This should allow single worker systems to impose a limit on the number of shard consumers without grinding to a halt.

### CODE REVIEW

File: src/main/java/com/fivetran/external/com/amazonaws/services/kinesis/clientlibrary/lib/worker/ParentsFirstShardPrioritization.java [hunk 1]
```diff
@@ -48,38 +48,38 @@ public ParentsFirstShardPrioritization(int maxDepth) {
...
+                orderedInfos.add((T) sortingTreeNode.shardInfo);
             }
         }
         return orderedInfos;
     }
 
-    private int populateDepth(String shardId,
-            Map<String, ShardInfo> shards,
-            Map<String, SortingNode> processedNodes) {
+    private <T extends PriorizeableShard> int populateDepth(String shardId,
```
Comment (bencvdb) [inline src/main/java/com/fivetran/external/com/amazonaws/services/kinesis/clientlibrary/lib/worker/ParentsFirstShardPrioritization.java:80]: Note this method is recursive. I tested it out and it handled a lineage of 4500 shards.

File: .gitignore [hunk 2]
```diff
@@ -1,3 +1,4 @@
 target/
 AwsCredentials.properties
 .idea
+*.iml
```
Comment (bencvdb) [inline .gitignore:4]: files generated by IntelliJ. Ignored in our `engineering` repo too.

**Review by zekail — APPROVED** [review]

File: src/main/java/com/fivetran/external/com/amazonaws/services/kinesis/clientlibrary/lib/worker/ParentsFirstShardPrioritization.java (hunk 1 above)
Comment (fivetran-alvinchang) [reply src/main/java/com/fivetran/external/com/amazonaws/services/kinesis/clientlibrary/lib/worker/ParentsFirstShardPrioritization.java:80]: If you are concerned that we may hit the recursion limit (seems unlikely) we could do this iteratively instead.

File: src/main/java/com/fivetran/external/com/amazonaws/services/kinesis/clientlibrary/lib/worker/PriorizeableShard.java [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package com.fivetran.external.com.amazonaws.services.kinesis.clientlibrary.lib.worker;
+
+import java.util.Collection;
+
+public interface PriorizeableShard {
```
Comment (fivetran-alvinchang) [inline src/main/java/com/fivetran/external/com/amazonaws/services/kinesis/clientlibrary/lib/worker/PriorizeableShard.java:5]: nit: sp(?) seems like in some places you use `Priorizeable` vs `Prioritizable`

File: src/main/java/com/fivetran/external/com/amazonaws/services/kinesis/leases/impl/LeaseFetchOrder.java [hunk 4]
```diff
@@ -0,0 +1,6 @@
+package com.fivetran.external.com.amazonaws.services.kinesis.leases.impl;
+
+public enum LeaseFetchOrder {
+    RANDOM,
```
Comment (fivetran-alvinchang) [inline src/main/java/com/fivetran/external/com/amazonaws/services/kinesis/leases/impl/LeaseFetchOrder.java:4]: Out of curiosity - Once we roll out the `PARENTS_FIRST` fetch order is there any benefit to keeping the RANDOM fetch order?

File: src/test/java/com/fivetran/external/com/amazonaws/services/kinesis/clientlibrary/lib/worker/ParentsFirstKinesisClientLeastPrioritzationUnitTest.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
+package com.fivetran.external.com.amazonaws.services.kinesis.clientlibrary.lib.worker;
+
+import com.fivetran.external.com.amazonaws.services.kinesis.leases.impl.KinesisClientLease;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+
+public class ParentsFirstKinesisClientLeastPrioritzationUnitTest extends ParentsFirstShardPrioritizationUnitTest<KinesisClientLease> {
```
Comment (fivetran-alvinchang) [inline src/test/java/com/fivetran/external/com/amazonaws/services/kinesis/clientlibrary/lib/worker/ParentsFirstKinesisClientLeastPrioritzationUnitTest.java:9]: nit: sp(?) ParentsFirstKinesisClientLeastPrioritzationUnitTest -> ParentsFirstKinesisClientLeastPriorit**i**zationUnitTest

File: src/test/java/com/fivetran/external/com/amazonaws/services/kinesis/clientlibrary/lib/worker/ParentsFirstShardInfoPrioritzationUnitTest.java [hunk 6]
```diff
@@ -0,0 +1,25 @@
+package com.fivetran.external.com.amazonaws.services.kinesis.clientlibrary.lib.worker;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public class ParentsFirstShardInfoPrioritzationUnitTest extends ParentsFirstShardPrioritizationUnitTest<ShardInfo> {
```
Comment (fivetran-alvinchang) [inline src/test/java/com/fivetran/external/com/amazonaws/services/kinesis/clientlibrary/lib/worker/ParentsFirstShardInfoPrioritzationUnitTest.java:7]: nit: sp(?)

File: src/test/java/com/fivetran/external/com/amazonaws/services/kinesis/clientlibrary/lib/worker/ParentsFirstKinesisClientLeastPrioritzationUnitTest.java (hunk 5 above)
Comment (fivetran-alvinchang) [reply src/test/java/com/fivetran/external/com/amazonaws/services/kinesis/clientlibrary/lib/worker/ParentsFirstKinesisClientLeastPrioritzationUnitTest.java:9]: I guess Google thinks `prioritizable` is a word but IntelliJ thinks it's a typo  ¯\_(ツ)_/¯

File: src/main/java/com/fivetran/external/com/amazonaws/services/kinesis/leases/impl/LeaseFetchOrder.java (hunk 4 above)
Comment (bencvdb) [reply src/main/java/com/fivetran/external/com/amazonaws/services/kinesis/leases/impl/LeaseFetchOrder.java:4]: Yes. This Worker was designed to be used in a multi-Worker setup where they all are pulling from the same pool of leases. Having them select from the pool randomly reduces the chances of collision. You can read more about the intended setup here: https://aws.amazon.com/blogs/big-data/process-large-dynamodb-streams-using-multiple-amazon-kinesis-client-library-kcl-workers/

I want to preserve this functionality for two reasons:
1. Make sure it is available to us if we dramatically rethink how to use the dynamo connector and use multiple workers per table
2. This repo is open-source and I felt like if someone else may want to have both options available.

File: src/test/java/com/fivetran/external/com/amazonaws/services/kinesis/clientlibrary/lib/worker/ParentsFirstShardInfoPrioritzationUnitTest.java (hunk 6 above)
Comment (bencvdb) [reply src/test/java/com/fivetran/external/com/amazonaws/services/kinesis/clientlibrary/lib/worker/ParentsFirstShardInfoPrioritzationUnitTest.java:7]: good... "I"

**Pull request closed without merging**